### DISCUSSION TITLE
Dev Module External Pages

### DISCUSSION THREAD

**Discussion by JSatterly** [oqtane/oqtane.framework#2675] [Q&A] [answered]
When a Developer Module has been built, it is set up as a separate solution with Client, Server, Package and Shared projects. Then the Octane Framework Server project is added and made the startup project.  There are Pages folders in both the Client and Server projects where, I'm guessing, either API or EF would be used for data connections. Is there a "best practice" on where external razor pages and code should go if the intent is still to run the final app server side? Is one or the other more in keeping with the Oqtane base functionality?

**Comment by sbwalker** [comment]
When you say "external razor pages" I assume you mean actual razor pages (.cshtml) and not razor components (.razor). Razor pages (.cshtml) are a server-side technology and would never be executed on the client. Oqtane only has razor pages within its Oqtane.Server project... in the /Pages folder which contains the _Host.cshtml main bootstrap page as well as a few other pages for non-UI processing (ie. for dealing with login/logout authentication). If you actually need to create your own razor pages (.cshtml) you should put them in your Server project and you should use the route "/pages" as a prefix (ie. @page "/pages/mypage") so that Oqtane handles them correctly. When Oqtane loads your Server assembly at startup it will inspect it to determine if it contains any razor pages, and if so, it will register them for you (see \Oqtane.Server\Extensions\OqtaneMvcBuilderExtensions.cs). What is the functionality you expect to include in your razor pages? Is it non-UI logic?

**Reply by ghost** [reply]
@sbwalker I have a .cshtml page in the external module's server project with route `@page "/pages/account/login"`. Its non-UI logic (setting some cookies after login).

However, Oqtane doesn't recognize my page as mentioned it should. Instead it redirects to `http://localhost:44357/login?returnurl=%2Fpages%2Faccount%2Flogin%2F` even though route is [AllowAnonymous]. 

I tried copying login page in oqtane project itself and it recognize that but not from external module. Any suggestion?

**Reply by sbwalker** [reply]
@farhanalie there is code in Oqtane.Server\Extensions\OqtaneMvcBuilderExtensions.cs which inspects the DLLs in the /bin and detects if they implement MVC or Razor Page types, and if so, registers them automatically. Note that the DLLs must follow the Oqtane naming convention or else they will not be inspected (ie. your external module Server assembly must be named "{yourmodulename}.Server.oqtane.dll")

**Reply by ghost** [reply]
@sbwalker thanks for the quick reply.
I double-checked, and DLLs are following the convention, In fact, I created a module with a creator from Oqtane. This is name of my server dll SimCore.UI.Server.Oqtane.dll, where I have a page with @page "/pages/account/login". For some reason, it's still not been able to recognize it as a post URL / endpoint. 
![image](https://github.com/oqtane/oqtane.framework/assets/21053342/116ffc61-01b2-4d84-8974-f6334e53a0f1)

**Reply by sbwalker** [reply]
@farhanalie does your Razor Page have a separate DLL for its Views (ie. *.Views.dll)?

**Reply by ghost** [reply]
@sbwalker  Nope. it should be part module server dll. 
SimCore.UI.Server.Oqtane.dll

**Reply by sbwalker** [reply]
I would suggest you put a breakpoint in the AddOqtaneApplicationParts() method in Oqtane and see what is happening when it inspects your assembly.

**Reply by sbwalker** [reply]
@farhanalie I just ran a test by adding a razor page to an Oqtane module created using the Module Creator. I created a Pages folder in the Server project and I added a Test.cshtml:
```
@page "/pages/test"
@namespace MyCompany.MyModule.Pages
@model MyCompany.MyModule.Pages.TestModel
```
and a Test.cshtml.cs
```
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.Mvc.RazorPages;

namespace MyCompany.MyModule.Pages
{
    [AllowAnonymous]
    public class TestModel : PageModel
    {
        public IActionResult OnGet()
        {
            return Content("Test");
        }
    }
}
```
I can see the MyCompany.MyModule.Server.Oqtane.dll assembly being processed by the AddOqtaneApplicationParts() method... and I am able to successfully access the url at /pages/test

**Reply by ghost** [reply]
thanks @sbwalker for the help your example was undoubtedly a starting point for me. The problem I had was the file name of Login.cshtml, Login.cshtml.cs and model name. So I updated both file name and model name. Even though namespace and assembly are different, I think it is somehow getting conflicted with Oqtane's login page. So updating name of file and model name worked for me.

**Comment by JSatterly** [comment]
Sorry, my bad use of terms, I do mean UI files. But that was my wonder - if I do want to add UI files but intend to run it as a server app rather than a hosted webassembly app, would these files still be added into the Client project and use the API for data rather than adding them in the Server project and using EF more directly for the data?

**Comment by sbwalker** [comment]
Yes it would be possible to create your razor components (.razor) in your Server project. Oqtane inspects all properly named assemblies for module interfaces during startup and will be able to find them in the Server project. 

The reason there are no examples of this approach is that it severely limits the use of your modules. They will only be able to function in the Blazor Server hosting model - not Blazor WebAssembly or Blazor Hybrid. Only a true client/server architecture supports all hosting models. 

If you want to take the server-only approach you may also run into DBContext issues due to the difference in scoping on Blazor Server. There are recommended approaches from Microsoft on how to deal with this - you would need to do some research.

**Comment by sbwalker** [comment, marked as answer]
If you only expect to ever run your application on Blazor Server then including your razor components in the Server project would significantly reduce the amount of code you need to write and maintain in the long term as you would not need Services/Controllers.

Another thread which discusses this topic is #1843 - it describes an alternative architecture which avoids using a REST API when running on Blazor Server - and uses the REST API when running on Blazor WebAssembly/Hybrid. This approach would not reduce the amount of code - it would actually result in more code (ie. more layers to develop and maintain) - but it does offer more flexibility. It also mentions some of the challenges which would be in common with your original question.

**Comment by JSatterly** [comment]
Perfect, thanks!
